### PR TITLE
Features/spi packet casts

### PR DESCRIPTION
**Pull request by ricardochust**
Changed the name for the SPIPackets to SPIOrders, as they now serve the purpose of the latter.

Created a new type of variable called SPIPacket, which serves to define the structure of the master or the slave data payload. 

Now there is a function called set_callback that lets the user set a function to be called when the Order is succesfully sent. 

Added a queue for the Orders on the SPI, using a new module called RingBuffer (does exactly what it names implies)

Added an optional RS GPIO pin that will signal the master when the slave is ready so he doesn t wait until the next check on the orders_update function (makes it faster)

There are now different types of SPIOrders, and each of them handles data on a different way: 
-SPIBinaryOrder works as the old SPIOrder did (called SPIPacket in that version).
-SPIWordOrder sends full words (uint32_t) of data allocated inside its master_array and slave_array
-SPIStackOrder is a replica of the functionality of the StackOrder on ethernet, but uses the SPIPacket variable to define the structure instead of having it defined in the template of the constructor.

### CODE REVIEW

**Comment by ricardochust** [comment]
Working with ethernet on. 

Will need a MPU_start() for no ethernet to work as it uses the cache.

File: Inc/HALAL/Models/Packets/SPIOrder.hpp [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+ * SpiPacket.hpp
+ *
+ *  Created on: 26 jan. 2024
+ *      Author: ricardo
+ */
+
+
+#include "SPIPacket.hpp"
+
+#define ALIGN_NUMBER_TO_32(x) (((x + 31)/32)*32)
```
Comment (g0nz4I0) [inline Inc/HALAL/Models/Packets/SPIOrder.hpp:11]: what does this do?

File: Inc/HALAL/Models/Packets/SPIPacket.hpp [hunk 2]
```diff
@@ -8,48 +8,79 @@
...
+/**
+ * @brief template class used as a structure to link variables of any type to one or more SPIPackets.
+ */
+template<size_t BufferLength, class... Types>
+class SPIPacket : public SPIBasePacket{
+public:
+    PacketValue<>* values[sizeof...(Types)];
+    uint8_t buffer[BufferLength];
+    stack_tuple<PacketValue<Types>...> packetvalue_warehouse;
+    size_t& size = SPIBasePacket::size;
```
Comment (g0nz4I0) [inline Inc/HALAL/Models/Packets/SPIPacket.hpp:43]: why the use of a reference?

File: Inc/HALAL/Services/Communication/SPI/SPI.hpp [hunk 3]
```diff
@@ -8,14 +8,17 @@
...
+#include "Packets/SPIOrder.hpp"
 #include "DigitalOutputService/DigitalOutputService.hpp"
+#include "DigitalInputService/DigitalInputService.hpp"
 #include "DMA/DMA.hpp"
 #include "ErrorHandler/ErrorHandler.hpp"
 
 #ifdef HAL_SPI_MODULE_ENABLED
 
-#define MASTER_SPI_CHECK_DELAY 100000 //how often the master should check if the slave is ready, in nanoseconds
+#define MASTER_SPI_CHECK_DELAY 400000 //how often the master should check if the slave is ready, in nanoseconds
```
Comment (g0nz4I0) [inline Inc/HALAL/Services/Communication/SPI/SPI.hpp:19]: nit pick, add NS suffix to macro, so if comment is ever lost variable holds info

File: Inc/HALAL/Services/Communication/SPI/SPI.hpp [hunk 4]
```diff
@@ -123,13 +138,25 @@ class SPI{
      */
     static uint8_t inscribe(SPI::Peripheral& spi);
 
+
+    /**
+     * @brief assigns the RS pin to the SPI.
+     */
+    static void assign_RS(uint8_t id, Pin& RSPin);
```
Comment (g0nz4I0) [inline Inc/HALAL/Services/Communication/SPI/SPI.hpp:145]: maybe assign_RS_pin() ?

File: Inc/HALAL/Services/Communication/SPI/SPI.hpp [hunk 5]
```diff
@@ -176,35 +203,101 @@ class SPI{
...
 	 * @param spi Id of the SPI
 	 */
     static void chip_select_off(uint8_t id);
 
+    static inline void spi_communicate_cache_data(SPI::Instance* spi, uint8_t* value_to_send, uint16_t size_to_send, uint8_t* value_to_receive, uint16_t aligned_size_to_receive);
+    static inline void spi_end_cache_data_communication(uint8_t* value_to_receive, uint16_t size_to_receive);
+
+
+    static void turn_on_chip_select(SPI::Instance* spi);
+    static void turn_off_chip_select(SPI::Instance* spi);
```
Comment (g0nz4I0) [inline Inc/HALAL/Services/Communication/SPI/SPI.hpp:274]: no me deja github seleccionar dos lineas no contiguas, pq hay un chip_select off y un turn_off_chip_select??

sobretodos los dos publicos, pq entiendo q uno sera a traves del id, pero el otro es a traves del instance

File: Src/HALAL/Services/ADC/ADC.cpp [hunk 6]
```diff
@@ -17,15 +17,15 @@ unordered_map<uint8_t, ADC::Instance> ADC::active_instances = {};
...
 	}
 
 	if(hadc == ADC::peripherals[1].handle){
-		SCB_CleanInvalidateDCache_by_Addr((uint32_t*)ADC::peripherals[1].dma_stream, 32);
+	//	SCB_CleanInvalidateDCache_by_Addr((uint32_t*)ADC::peripherals[1].dma_stream, 32);
 	}
 	
 	if(hadc == ADC::peripherals[2].handle){
-		SCB_CleanInvalidateDCache_by_Addr((uint32_t*)ADC::peripherals[2].dma_stream, 32);
+	//	SCB_CleanInvalidateDCache_by_Addr((uint32_t*)ADC::peripherals[2].dma_stream, 32);
```
Comment (g0nz4I0) [inline Src/HALAL/Services/ADC/ADC.cpp:28]: remove the code completely, no code should ever be pushed commented out.

File: Src/HALAL/Services/Communication/SPI/SPI.cpp [hunk 7]
```diff
@@ -128,45 +160,52 @@ bool SPI::transmit_and_receive(uint8_t id, span<uint8_t> command_data, span<uint
...
+		return spi->SPIOrderQueue.push(Order.id);
 	}
 
-	spi->state = SPI::STARTING_PACKET;
-	spi->SPIPacketID = packet.id;
-	SPI::chip_select_off(id);
-	HAL_SPI_TransmitReceive_DMA(spi->hspi, (uint8_t *)&spi->SPIPacketID, (uint8_t *)&spi->available_end, 2);
+	spi->state = SPI::STARTING_ORDER;
+	*(spi->SPIOrderID) = Order.id;
+	master_check_available_end(spi);
```
Comment (g0nz4I0) [inline Src/HALAL/Services/Communication/SPI/SPI.cpp:182]: i dont like that this indirectly sends the message :thinking:

File: Src/HALAL/Services/Communication/SPI/SPI.cpp [hunk 8]
```diff
@@ -128,45 +160,52 @@ bool SPI::transmit_and_receive(uint8_t id, span<uint8_t> command_data, span<uint
...
+	SPI::Instance* spi = SPI::registered_spi[id];
+
+	if(spi->state != SPI::IDLE || !spi->SPIOrderQueue.is_empty()){
+		return spi->SPIOrderQueue.push(Order->id);
 	}
+
+	spi->state = SPI::STARTING_ORDER;
+	*(spi->SPIOrderID) = Order->id;
+	master_check_available_end(spi);
+	//SPI::chip_select_off(id);
```
Comment (g0nz4I0) [inline Src/HALAL/Services/Communication/SPI/SPI.cpp:202]: remove this commented code

File: Src/HALAL/Services/Communication/SPI/SPI.cpp [hunk 9]
```diff
@@ -317,23 +452,75 @@ void HAL_SPI_RxCpltCallback(SPI_HandleTypeDef *hspi){
...
+#ifdef STLIB_ETH
+	SCB_CleanDCache_by_Addr((uint32_t*)value_to_send, size_to_send);
+	SCB_InvalidateDCache_by_Addr(value_to_receive, aligned_size_to_receive);
+#endif
+	HAL_SPI_TransmitReceive_DMA(spi->hspi, value_to_send, value_to_receive, size_to_send);
+}
+
+
+void SPI::spi_end_cache_data_communication(uint8_t* value_to_receive, uint16_t size_to_receive){
+#ifdef STLIB_ETH
```
Comment (g0nz4I0) [inline Src/HALAL/Services/Communication/SPI/SPI.cpp:475]: as soon as the MPU branch is merged I'm guessing we will change this right?

Comment (ricardochust) [reply Src/HALAL/Services/Communication/SPI/SPI.cpp:475]: indeed

**Pull request closed without merging**